File: Python/4-introspection.py
"""
Introspection examples. Using my_ before functions to
not redefine built-in functions.
"""

import inspect


def my_sum(a, b):
    return a + b


my_max = lambda x, y: x if x > y else y

print('Names:')
print('my_sum:', my_sum.__name__)
print('my_max:', my_max.__name__)

print('Arguments:')
print('my_sum:', str(inspect.signature(my_sum)))
print('my_max:', str(inspect.signature(my_max)))

print('Anonymous function: ' + (lambda x: x).__name__)

print('Function code:')
print('my_sum:', inspect.getsource(my_sum))
print('my_max:', inspect.getsource(my_max))

File: JavaScript/4-introspection.js
'use strict';

function inc(a) {
  return ++a;
}

const sum = function(a, b) {
  return a + b;
};

const max = (a, b) => (a > b ? a : b);

console.log('Names: ');
console.dir({
  inc: inc.name,
  sum: sum.name,
  max: max.name,
});

console.log('Arguments: ');
console.dir({
  inc: inc.length,
  sum: sum.length,
  max: max.length,
});

console.log('Anonymous function: ' + function(x) { return x; }.name);
console.log('Anonymous lambda: ' + ((x) => x).name);

console.log('toString: ');
console.dir({
  inc: inc.toString(),
  sum: sum.toString(),
  max: max.toString(),
});
